src/ripple/app/misc/impl/Manifest.cpp

```diff
@@ -27,44 +27,129 @@
 #include <ripple/json/json_reader.h>
 #include <ripple/protocol/PublicKey.h>
 #include <ripple/protocol/Sign.h>
+#include <boost/algorithm/clamp.hpp>
 #include <boost/regex.hpp>
 #include <numeric>
 #include <stdexcept>
 
 namespace ripple {
 
-boost::optional<Manifest>
-Manifest::make_Manifest (std::string s)
+boost::optional<Manifest> deserializeManifest(Slice s)
 {
+    if (s.empty())
+        return boost::none;
+
+    static SOTemplate const manifestFormat (
+        [](SOTemplate& t)
+        {
+            // A manifest must include:
+            // - master public key:
+            t.push_back (SOElement (sfPublicKey,       SOE_REQUIRED));
+
+            // - signature with that public key:
+            t.push_back (SOElement (sfMasterSignature, SOE_REQUIRED));
+
+            // - sequence number:
+            t.push_back (SOElement (sfSequence,        SOE_REQUIRED));
+
+            // It may, optionally, contain:
+            // - a version number which defaults to 0:
+            t.push_back (SOElement (sfVersion,          SOE_DEFAULT));
+
+            // - a domain name
+            t.push_back (SOElement (sfDomain,           SOE_OPTIONAL));
+
+            // - an ephemeral signing key that can be changed as necessary
+            t.push_back (SOElement (sfSigningPubKey,    SOE_OPTIONAL));
+
+            // - an signature using the ephemeral signing key, if it is present:
+            t.push_back (SOElement (sfSignature,        SOE_OPTIONAL));
+        });
+
     try
     {
-        STObject st (sfGeneric);
-        SerialIter sit (s.data (), s.size ());
-        st.set (sit);
+        SerialIter sit{ s };
+        STObject st{ sit, sfGeneric };
+
+        st.applyTemplate(manifestFormat);
+
+        // We only understand "version 0" manifests at this time:
```

Comment:
By adding a new field, is this not a "new version" of manifests? Or do the versions only count (backwards?/forwards?)-incompatible changes?

Reply:
The `version` field is designed in such a way so as to be compatible: manifests without the field are, implicitly, version 0. Explicitly specifying the value 0 in this field is wrong. Future manifest enhancements that necessitate a bump in the version number will include the field set to 1, 2, etc.